src/storage/StorageHttpDownloadHandler.cpp

```diff
@@ -0,0 +1,171 @@
+     }
+     hdfsUrl_ = headers->getQueryParam("url");
+     hdfsPort_ = headers->getIntQueryParam("port");
+     partitions_ = headers->getQueryParam("parts");
+     hdfsPath_ = headers->getQueryParam("path");
+     localPath_ = headers->getQueryParam("local");
+}
+
+
+void StorageHttpDownloadHandler::onBody(std::unique_ptr<folly::IOBuf>) noexcept {
+    // Do nothing, we only support GET
+}
+
+
+void StorageHttpDownloadHandler::onEOM() noexcept {
+    switch (err_) {
+        case HttpCode::E_UNSUPPORTED_METHOD:
+            ResponseBuilder(downstream_)
+                .status(405, "Method Not Allowed")
+                .sendWithEOM();
+            return;
+        case HttpCode::E_ILLEGAL_ARGUMENT:
+            ResponseBuilder(downstream_)
+                .status(400, "Bad Request")
+                .sendWithEOM();
+            return;
+        default:
+            break;
+    }
+
+    if (auto hadoopHome = std::getenv("HADOOP_HOME")) {
+        LOG(INFO) << "Hadoop Path : " << hadoopHome;
+        std::vector<std::string> parts;
+        folly::split(",", partitions_, parts, true);
+        if (downloadSSTFiles(hdfsUrl_, hdfsPort_, hdfsPath_, parts, localPath_)) {
+            ResponseBuilder(downstream_)
+                .status(200, "SSTFile download successfully")
+                .body("SSTFile download successfully")
+                .sendWithEOM();
+        } else {
+            ResponseBuilder(downstream_)
+                .status(404, "SSTFile download failed")
+                .sendWithEOM();
+        }
+    } else {
+        ResponseBuilder(downstream_)
+            .status(404, "HADOOP_HOME not exist")
+            .sendWithEOM();
+    }
+}
+
+
+void StorageHttpDownloadHandler::onUpgrade(UpgradeProtocol) noexcept {
+    // Do nothing
+}
+
+
+void StorageHttpDownloadHandler::requestComplete() noexcept {
+    delete this;
+}
+
+
+void StorageHttpDownloadHandler::onError(ProxygenError error) noexcept {
+    LOG(ERROR) << "Web service StorageHttpDownloadHandler got error: "
+               << proxygen::getErrorString(error);
+}
+
+std::string StorageHttpDownloadHandler::toStr(folly::dynamic& vals) const {
+    std::stringstream ss;
+    for (auto& counter : vals) {
+        auto& val = counter["value"];
+        ss << counter["name"].asString() << "="
+           << val.asString()
+           << "\n";
+    }
+    return ss.str();
+}
+
+bool StorageHttpDownloadHandler::downloadSSTFiles(const std::string& url,
+                                          int port,
+                                          const std::string& path,
+                                          const std::vector<std::string>& parts,
+                                          const std::string& local) {
+    static folly::IOThreadPoolExecutor executor(FLAGS_download_thread_num);
+    std::condition_variable cv;
+    std::mutex lock;
+    std::atomic<int> completed(0);
+    std::atomic<bool> successful{true};
+
+    for (auto& part : parts) {
```

Comment:
Add a flag to control there is only one download request could be invoked at the same time.